design-proposals/preferences.md

```diff
@@ -0,0 +1,584 @@
+
+  // PreferredDedicatedIoThread optionally enables dedicated IO threads for Disk devices.
+  //
+  // +optional
+  PreferredDedicatedIoThread *bool `json:"preferredDedicatedIoThread,omitempty"`
+
+  // PreferredCache optionally defines the DriverCache to be used by Disk devices.
+  //
+  // +optional
+  PreferredCache v1.DriverCache `json:"preferredCache,omitempty"`
+
+  // PreferredIo optionally defines the QEMU disk IO mode to be used by Disk devices.
+  //
+  // +optional
+  PreferredIo v1.DriverIO `json:"preferredIo,omitempty"`
+
+  // PreferredBlockSize optionally defines the block size of Disk devices.
+  //
+  // +optional
+  PreferredBlockSize *v1.BlockSize `json:"preferredBlockSize,omitempty"`
+
+}
+
+// InputDefaults contains optional defaults for Inputs such as the preferred input model.
+//
+// +k8s:openapi-gen=true
+type InputDefaults struct {
+  // PreferredInputModel specifies the model for Input devices.
+  //
+  // +optional
+  PreferredInputModel string `json:"preferredInputModel,omitempty"`
+}
+
+// InterfaceDefaults contains optional defaults for Interfaces such as the preferred model to be used.
+//
+// +k8s:openapi-gen=true
+type InterfaceDefaults struct {
+
+  // PreferredInterfaceModel optionally defines the preferred model to be used by Interface devices.
+  //
+  // +optional
+  PreferredInterfaceModel string `json:"preferredInterfaceModel,omitempty"`
+}
+```
+
+## [`VirtualMachineFlavorProfile`](http://kubevirt.io/api-reference/main/definitions.html#_v1alpha1_virtualmachineflavorprofile)
+
+[`VirtualMachineFlavorProfile`](http://kubevirt.io/api-reference/main/definitions.html#_v1alpha1_virtualmachineflavorprofile) will be extended to include a [`Resource`](http://kubevirt.io/api-reference/main/definitions.html#_k8s_io_api_core_v1_resourcerequirements) attribute as found on the [DomainSpec](http://kubevirt.io/api-reference/main/definitions.html#_v1_domainspec) that can be used to request `CPU` and `Memory` resources. As when defining a standard `VMI` both a `CPU` topology and `Resources` request for `CPU` resources cannot be made in the same [`VirtualMachineFlavorProfile`](http://kubevirt.io/api-reference/main/definitions.html#_v1alpha1_virtualmachineflavorprofile).
+
+[`VirtualMachineFlavorProfile`](http://kubevirt.io/api-reference/main/definitions.html#_v1alpha1_virtualmachineflavorprofile) will also be extended to include an optional reference to a `PreferencesMatcher` to link the flavor to a specific set of preferences.
+
+```go
+// VirtualMachineFlavorProfile contains definitions that will be applied to VirtualMachine.
+//
+// +k8s:openapi-gen=true
+type VirtualMachineFlavorProfile struct {
+
+  // Name specifies the name of this custom profile.
+  Name string `json:"name"`
+
+  // Default specifies if this VirtualMachinePreferencesProfile is the default for the VirtualMachineFlavor.
+  // Zero or one profile can be set to default.
+  //
+  // +optional
+  Default bool `json:"default,omitempty"`
```

Comment:
Does this mean the whole API call is invalid, if multiple items with Default=true are sent or received?

Reply:
Yes, however I've dropped profiles from the proposal moving forward.